### DISCUSSION TITLE
Robot Toolkit in Grasshopper

### DISCUSSION THREAD

**Discussion by tuahhamid** [BHoM/BHoM#1567] [Q&A]
Hello,

Grasshopper and Robot newbie here. I have a Robot file that contains 4 bars (beam elements from Revit). Using  Robot adapter, I managed to capture the bar object from Robot to Grasshopper. What I'm trying to extract is the Length property of each bar object.

![Robot Length Properties](https://github.com/BHoM/BHoM/assets/88237170/ecf4f905-5ba6-457e-9f02-6571fd9b1215)

I tried exploding the object multiple times but I can't find any Length property. The current workaround I have at the moment is to get the EndNode and StartNode points and get the absolute values from subtracting both points but this is not ideal since real projects will have diagonal beams.
![gh_explode bar](https://github.com/BHoM/BHoM/assets/88237170/7db3f39d-d556-4c34-86a8-04576ac41d2a)

**Comment by FraserGreenroyd** [comment]
Hi @Tuah614 , welcome to BHoM 😄 

(tl;dr at the bottom 😉 )

One of BHoM's guiding principles is to allow the object model to be universally understood (as best we can) between the different disciplines and domains within the AEC industry, with the aim to [standardise the data](https://bhom.xyz/documentation/) we're using when undertaking design engineering. As part of that, we've deliberately kept the objects as simple as possible, containing only the properties which are necessary to define them for use - this does mean that some objects may appear to have less 'data' than perhaps they ought to, but by keeping them minimal it allows everyone to (hopefully) have a rich conversation around the object and its use, as opposed to debating whether the measurement property should be `length` or `width` or `depth` depending on your point of view.

That said, obviously the `length` of a bar is an important characteristic in many calculations, so it does exist, just not as a property of the bar. Instead, we use the [Engine Methods](https://bhom.xyz/documentation/BHoM_Engine/) to provide functionality on the objects and provide properties which can be obtained with reasonably easy computation on the data the objects contain.

This also helps us maintain the objects during operation. If we stored both the `length` and the geometry of the bar on the same object, if we ever later changed the geometry, we'd have to remember to update the `length` property or risk the two properties being out of sync and causing problems down the line. Thus, keeping the properties of the object solely to those which are necessary to define the object, and allowing everything else to be calculated helps keep the data entwined correctly.

Part of this, includes our [Query Methods](https://bhom.xyz/documentation/BHoM_Engine/BHoM_Engine-Classes/#query) framework within the engines, which exist to ask simple(ish) questions of the objects which can be derived easily from their existing properties. In this situation, the `length` query method for a `bar` object provides the calculated length from the geometry of the bar itself, allowing us to always get the correct length regardless of the geometry changes the bar might go through during its life span.

If you do a `ctrl` + `shift` + `b` command on the Grasshopper canvas, you can search for `length` and will likely get a few results show up, such as those in the image below, of which one can be used to complete the query on the bar. Alternatively, you can also find it by dragging out from the bar component and releasing the mouse, which should bring up a search window in which you can type `length` and it will filter out to the methods which would accept the bar as an input.

Screenshot showing the full `length` search list:
![image](https://github.com/BHoM/BHoM/assets/18049174/e2bac095-42a4-4486-a251-179cc470c6c6)

Screenshot showing the `length` search with `bar` as a first method input:
![image](https://github.com/BHoM/BHoM/assets/18049174/d43f9da0-b8e0-45c3-987e-1395d18c19b6)

For your particular query of the `bar` itself, I would recommend the [Spatial Engine Length query method](https://github.com/BHoM/BHoM_Engine/blob/develop/Spatial_Engine/Query/Length.cs#L44) which takes in a 1D Element and returns its length in SI units (see below).

![image](https://github.com/BHoM/BHoM/assets/18049174/fc78faee-d638-4598-b9c3-72872950bd8f)

I recognise that I can easily answer this question because I've been involved in BHoM for the best part of 6 years, helping shape it on the Environments side and the DevOps side in that time, which gives me a very unfair advantage on knowing where things are (such as knowing I can use the Spatial Engine query method on a Bar for example). Therefore, any suggestions you have for improving the way we find things in BHoM will be greatly appreciated and a welcome addition to our open source community - we want a tool that everyone can use and can only achieve that with the support of everyone using it 😄 I have a few things lined up for the 2024 development calendar to improve UI/UX usage, so if you do have any suggestions please do get in touch! 🚀 

***

tl;dr: the `length` property is a calculated property of a `bar` so needs to use a query method to obtain, of which the Spatial Engine Length method is the best placed option for a `bar` currently 😄

**Comment by al-fisher** [comment]
Hi @Tuah614 - thanks for raising the question.

The _Length_ of a _Bar_ can be obtained through a "**Query**" method. 
When you explode a BHoM object you get access to all of its *Defining* properties. 
Any property like Length for example that is _Derived_ or calculated from Defining properties (i.e. like Length calculated from the Start and End Nodes) can be accessed by a Query. 

There are a number of ways to access Query methods through the menus and search.

![image](https://github.com/BHoM/BHoM/assets/15924573/c9b50d26-e89c-4104-a1b6-3abbc50b4aa3)

![image](https://github.com/BHoM/BHoM/assets/15924573/19728f3f-2c2b-447b-af04-b302f64d3351)

One nice quick way to get the right method for your Bar object input is to drag the wire from the Bar, release and then search for Length. See below.

![QueryLengthFromBar](https://github.com/BHoM/BHoM/assets/15924573/da77fc04-a27f-4765-a685-5ad8552e87ff)


One other thing to note - the above approach helps you find the right Length query. A Bar is an IElement1D - a linear line/curve based element. So you want the Length(IElement1D) as above.

![image](https://github.com/BHoM/BHoM/assets/15924573/59b8235d-f94c-4bd3-ae22-8c0d369774fc)

Hope this helps! 😄

**Comment by tuahhamid** [comment]
Wow @FraserGreenroyd  @al-fisher I didn't expect such detailed explanation! :green_heart: I appreciate the extra context you provided on BHoM's design principle - super useful especially for beginners like me who doesn't really know where to look for things.

**Reply by al-fisher** [reply]
You are most welcome! Really glad our responses were helpful!
